pkg/segment/metadata/metadata_test.go

```diff
@@ -0,0 +1,179 @@
+		nextVal--
+	}
+	isTableSorted := sort.SliceIsSorted(tableSorted, func(i, j int) bool {
+		return tableSorted[i].LatestEpochMS > tableSorted[j].LatestEpochMS
+	})
+	assert.True(t, isTableSorted)
+
+	isAllDataSorted := sort.SliceIsSorted(GetAllSegmentMicroIndexForTest(), func(i, j int) bool {
+		return GetAllSegmentMicroIndexForTest()[i].LatestEpochMS > GetAllSegmentMicroIndexForTest()[j].LatestEpochMS
+	})
+	assert.True(t, isAllDataSorted)
+}
+
+func testBlockRebalance(t *testing.T, fileCount int) {
+	blockSizeToLoad := GetAllSegmentMicroIndexForTest()[0].MicroIndexSize + GetAllSegmentMicroIndexForTest()[0].SearchMetadataSize
+	RebalanceInMemoryCmi(blockSizeToLoad)
+
+	assert.True(t, GetAllSegmentMicroIndexForTest()[0].AreMicroIndicesLoaded())
+	loadedBlockFile := GetAllSegmentMicroIndexForTest()[0].SegmentKey
+	loadedBlockLatestTime := GetAllSegmentMicroIndexForTest()[0].LatestEpochMS
+	assert.True(t, GetSegmentMetadataReverseIndexForTest()[loadedBlockFile].AreMicroIndicesLoaded())
+	assert.Same(t, GetAllSegmentMicroIndexForTest()[0], GetSegmentMetadataReverseIndexForTest()[loadedBlockFile])
+
+	for i := 1; i < fileCount; i++ {
+		assert.False(t, GetAllSegmentMicroIndexForTest()[i].AreMicroIndicesLoaded())
+		currFile := GetAllSegmentMicroIndexForTest()[i].SegmentKey
+		assert.Same(t, GetAllSegmentMicroIndexForTest()[i], GetSegmentMetadataReverseIndexForTest()[currFile])
+		assert.LessOrEqual(t, GetAllSegmentMicroIndexForTest()[i].LatestEpochMS, loadedBlockLatestTime, "we loaded the lastest timestamp")
+	}
+
+	// load just one more search metadata
+	blockSizeToLoad += GetAllSegmentMicroIndexForTest()[1].SearchMetadataSize
+	RebalanceInMemoryCmi(blockSizeToLoad)
+	assert.True(t, GetAllSegmentMicroIndexForTest()[0].AreMicroIndicesLoaded())
+	assert.False(t, GetAllSegmentMicroIndexForTest()[1].AreMicroIndicesLoaded(), "only load search metadata for idx 1")
+	for i := 2; i < fileCount; i++ {
+		assert.False(t, GetAllSegmentMicroIndexForTest()[i].AreMicroIndicesLoaded())
+		assert.False(t, GetAllSegmentMicroIndexForTest()[i].isSearchMetadataLoaded())
+		currFile := GetAllSegmentMicroIndexForTest()[i].SegmentKey
+		assert.Same(t, GetAllSegmentMicroIndexForTest()[i], GetSegmentMetadataReverseIndexForTest()[currFile])
+		assert.LessOrEqual(t, GetAllSegmentMicroIndexForTest()[i].LatestEpochMS, loadedBlockLatestTime, "we loaded the lastest timestamp")
+	}
+}
+
+func Test_RebalanceMetadata(t *testing.T) {
+	_ = localstorage.InitLocalStorage()
+	ResetGlobalMetadataForTest()
+
+	fileCount := 3
+	createMockMetaStore("data/", fileCount)
+	assert.Len(t, GetAllSegmentMicroIndexForTest(), fileCount)
+	assert.Len(t, GetSegmentMetadataReverseIndexForTest(), fileCount)
+
+	RebalanceInMemoryCmi(0)
+	for i := 0; i < fileCount; i++ {
+		assert.False(t, GetAllSegmentMicroIndexForTest()[i].AreMicroIndicesLoaded())
+	}
+
+	testBlockRebalance(t, fileCount)
+
+	err := os.RemoveAll("data/")
+	if err != nil {
+		assert.FailNow(t, "failed to remove data %+v", err)
+	}
```

Comment:
nit: if this test panics before here (or if the test is changed and it exits before here, e.g., with a bad `require.True()`), then this cleanup code won't run. You can instead put this at the start of the function (and add error handling if you want):
```
	t.Cleanup(func() { os.RemoveAll("data/") })
```

Reply:
yup makes sense